program/src/processor.rs

```diff
@@ -175,34 +198,223 @@ fn process_create_proposal(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
+        // Ensure the proposal account is owned by the Paladin Governance program.
+        if proposal_info.owner != program_id {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the proposal account has enough space.
+        if proposal_info.data_len() != std::mem::size_of::<Proposal>() {
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        // Ensure the proposal account is not initialized.
+        if &proposal_info.try_borrow_data()?[0..8] != ArrayDiscriminator::UNINITIALIZED.as_slice() {
+            return Err(ProgramError::AccountAlreadyInitialized);
+        }
+
+        let clock = <Clock as Sysvar>::get()?;
+        let creation_timestamp = clock.unix_timestamp;
+
+        // Write the data.
+        let mut proposal_data = proposal_info.try_borrow_mut_data()?;
+        *bytemuck::try_from_bytes_mut::<Proposal>(&mut proposal_data)
+            .map_err(|_| ProgramError::InvalidAccountData)? = Proposal::new(
+            stake_authority_info.key,
+            creation_timestamp,
+            governance_config,
+        );
     }
 
-    // Ensure the proposal account has enough space.
-    if proposal_info.data_len() != std::mem::size_of::<Proposal>() {
-        return Err(ProgramError::InvalidAccountData);
+    // Initialize the proposal transaction account.
+    {
+        let (proposal_transaction_address, signer_bump_seed) =
+            get_proposal_transaction_address_and_bump_seed(proposal_info.key, program_id);
+        let bump_seed = [signer_bump_seed];
+        let proposal_transaction_signer_seeds =
+            collect_proposal_transaction_signer_seeds(proposal_info.key, &bump_seed);
+
+        // Ensure the provided proposal transaction address is the correct
+        // address derived from the program.
+        if !proposal_transaction_info
+            .key
+            .eq(&proposal_transaction_address)
+        {
+            return Err(PaladinGovernanceError::IncorrectProposalTransactionAddress.into());
+        }
+
+        // Ensure the proposal transaction account has not already been
+        // initialized.
+        if proposal_transaction_info.data_len() != 0 {
+            return Err(ProgramError::AccountAlreadyInitialized);
+        }
+
+        let state = ProposalTransaction::default();
+        let space = get_instance_packed_len(&state)? as u64;
+
+        // Allocate & assign.
+        invoke_signed(
+            &system_instruction::allocate(&proposal_transaction_address, space),
+            &[proposal_transaction_info.clone()],
+            &[&proposal_transaction_signer_seeds],
+        )?;
+        invoke_signed(
+            &system_instruction::assign(&proposal_transaction_address, program_id),
+            &[proposal_transaction_info.clone()],
+            &[&proposal_transaction_signer_seeds],
+        )?;
+
+        // Write the data.
+        state.serialize(&mut &mut proposal_transaction_info.try_borrow_mut_data()?[..])?;
```

Comment:
I can't remember if this is actually ok, but to be totally safe, can this use the `borsh::to_writer` to make sure we're creating a new slice rather than consuming the underlying one in the account info? As done at https://github.com/solana-labs/solana-program-library/blob/2d2c4bba858ecd4004de09bfed1c043896a6d12f/stake-pool/program/src/processor.rs#L917-L920

Reply:
Who do you think made those changes? :wink: